crates/fluvio-protocol-derive/src/util.rs

```diff
@@ -1,75 +1,256 @@
+
+                match attr_name.as_str() {
+                    $(
+                        $field if $opt.is_none() => {
+                            return Ok($block)
+                        },
+                        $field => {return Err($meta.error(concat!($attr_ident, " attribute already specified")))}
+                    )*
+
+                    _ => return Err($meta.error(concat!("unrecognized ", $attr_ident, " attribute"))),
+                }
+            })?;
         }
-    })
+    };
 }
 
-pub(crate) fn find_name_attribute<'a>(meta: &'a Meta, name: &str) -> Option<&'a MetaNameValue> {
-    find_meta(meta, name).map(|meta| match meta {
-        Meta::NameValue(name_value) => name_value,
-        _ => panic!("should not happen"),
-    })
+// This piece of code can be included in the macro but there are some issues
+// For example if we return this from the parse_attributes! macro with closure function
+// It works but there is an issue with clippy giving a "warning: try not to call a closure in the expression where it is declared"
+// Here is how this can be fixed if used in macro: https://stackoverflow.com/questions/76989679/how-to-call-a-closure-from-a-macro
+// So to do this in a less work around way we use a separate function like so:
+// We return $meta variable like shown here: https://github.com/rust-lang/rust-clippy/issues/1553
+pub fn parse_attributes_data(meta: &ParseNestedMeta) -> (Option<syn::Expr>, Span, String) {
+    // we can safely unwarp as this is already checked from the parse_attributes macro
+    let ident = meta.path.get_ident().unwrap();
+    // to_string so we can return the name of the ident
+    // helps to reduce the repetitiveness of attribute names
+    let attr_name = ident.to_string();
+    let attr_span = ident.span();
+
+    if let Ok(value) = meta.value() {
+        if let Ok(expr) = value.parse() {
+            return (Some(expr), attr_span, attr_name);
+        }
+    }
+
+    (None, attr_span, attr_name)
 }
 
-pub(crate) fn find_meta<'a>(meta: &'a Meta, name: &str) -> Option<&'a Meta> {
-    if let Meta::List(list) = meta {
-        for attr in list.nested.iter() {
-            if let NestedMeta::Meta(named_meta) = attr {
-                match named_meta {
-                    Meta::NameValue(meta_name_value) => {
-                        if meta_name_value.path.is_ident(name) {
-                            return Some(named_meta);
-                        }
-                    }
-                    Meta::Path(path) => {
-                        if path.is_ident(name) {
-                            return Some(named_meta);
-                        }
+pub(crate) use parse_attributes;
+
+pub fn get_expr_value_from_meta(meta: &ParseNestedMeta) -> syn::Result<PropAttrsType> {
+    let (expr, attr_span, attr_name) = parse_attributes_data(meta);
+    get_expr_value(&attr_name, &expr, attr_span)
+}
+pub fn get_expr_value<'a>(
+    attr_name: &'a str,
+    field: &'a Option<Expr>,
```

Comment:
Here and below `&'a Option<Expr>` should be replaced by `Option<&'a Expr>`. This will save from unnecessary cloning of `Exp`.